File: go/0155.min-stack/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:34
// leetgo: dev
// https://leetcode.com/problems/min-stack/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

type MinStack struct {
	values    []int
	top       int
	minValues []int
}

func Constructor() MinStack {
	return MinStack{
		values:    make([]int, 0),
		top:       -1,
		minValues: make([]int, 0),
	}
}

//lint:file-ignore ST1006 leetcode use "this"
func (this *MinStack) Push(val int) {
	if this.top == -1 {
		this.values = append(this.values, val)
		this.minValues = append(this.minValues, val)
		this.top += 1
	} else {
		this.values = append(this.values, val)
		this.top += 1

		if val < this.GetMin() {
			this.minValues = append(this.minValues, val)
		} else {
			this.minValues = append(this.minValues, this.minValues[len(this.minValues)-1])
		}
	}
}

func (this *MinStack) Pop() {
	this.values = this.values[:len(this.values)-1]
	this.minValues = this.minValues[:len(this.minValues)-1]
	this.top -= 1
}

func (this *MinStack) Top() int {
	return this.values[len(this.values)-1]
}

func (this *MinStack) GetMin() int {
	return this.minValues[len(this.minValues)-1]
}

func (this *MinStack) Display() {
	fmt.Println(this.values)
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	ops := Deserialize[[]string](ReadLine(stdin))
	params := MustSplitArray(ReadLine(stdin))
	output := make([]string, 0, len(ops))
	output = append(output, "null")

	obj := Constructor()

	for i := 1; i < len(ops); i++ {
		switch ops[i] {
		case "push":
			methodParams := MustSplitArray(params[i])
			val := Deserialize[int](methodParams[0])
			obj.Push(val)
			output = append(output, "null")
		case "pop":
			obj.Pop()
			output = append(output, "null")
		case "top":
			ans := Serialize(obj.Top())
			output = append(output, ans)
		case "getMin":
			ans := Serialize(obj.GetMin())
			output = append(output, ans)
		}
	}
	fmt.Println("\noutput:", JoinArray(output))
}

File: rust/src/0155.min-stack/solution.rs
// Created by Olgierd Palasz at 2026/08/04 21:20
// leetgo: dev
// https://leetcode.com/problems/min-stack/

use anyhow::Result;
use leetgo_rs::*;

// @lc code=begin

struct Data {
    value: i32,
    min: i32,
}

struct MinStack {
    stack: Vec<Data>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {
    fn new() -> Self {
        MinStack {
            stack: Vec::<Data>::new(),
        }
    }

    fn push(&mut self, value: i32) {
        if self.stack.is_empty() {
            self.stack.push(Data { value, min: value });
        } else if value < Self::get_min(self) {
            self.stack.push(Data { value, min: value });
        } else {
            self.stack.push(Data {
                value,
                min: Self::get_min(self),
            });
        }
    }

    fn pop(&mut self) {
        self.stack.pop();
    }

    fn top(&self) -> i32 {
        self.stack.last().unwrap().value
    }

    fn get_min(&self) -> i32 {
        self.stack.last().unwrap().min
    }
}

// @lc code=end

fn main() -> Result<()> {
    let ops: Vec<String> = deserialize(&read_line()?)?;
    let params = split_array(&read_line()?)?;
    let mut output = Vec::with_capacity(ops.len());
    output.push("null".to_string());

    #[allow(unused_mut)]
    let mut obj = MinStack::new();

    for i in 1..ops.len() {
        match ops[i].as_str() {
            "push" => {
                let method_params = split_array(&params[i])?;
                let value: i32 = deserialize(&method_params[0])?;
                obj.push(value);
                output.push("null".to_string());
            }
            "pop" => {
                obj.pop();
                output.push("null".to_string());
            }
            "top" => {
                let ans: i32 = obj.top().into();
                output.push(serialize(ans)?);
            }
            "getMin" => {
                let ans: i32 = obj.get_min().into();
                output.push(serialize(ans)?);
            }
            _ => panic!("unknown op"),
        }
    }

    println!("\noutput: {}", join_array(output));
    Ok(())
}
